clients/src/main/java/org/apache/kafka/common/requests/ListOffsetRequest.java

```diff
@@ -64,66 +73,117 @@ public PartitionData(long timestamp, int maxNumOffsets) {
         }
     }
 
+    /**
+     * Constructor for ListOffsetRequest v0
+     */
+    @Deprecated
     public ListOffsetRequest(Map<TopicPartition, PartitionData> offsetData) {
-        this(CONSUMER_REPLICA_ID, offsetData);
+        this(CONSUMER_REPLICA_ID, offsetData, 0);
     }
 
+    /**
+     * Constructor for ListOffsetRequest v0
+     */
+    @Deprecated
     public ListOffsetRequest(int replicaId, Map<TopicPartition, PartitionData> offsetData) {
-        super(new Struct(CURRENT_SCHEMA));
-        Map<String, Map<Integer, PartitionData>> topicsData = CollectionUtils.groupDataByTopic(offsetData);
+        this(replicaId, offsetData, 0);
+    }
+
+    /**
+     * Constructor with a specified version.
+     */
+    public ListOffsetRequest(Map<TopicPartition, ?> targetTimes, int version) {
```

Comment:
I guess type erasure is making this tricky. Even so, the wildcard type seems less than ideal, right? Maybe we should just use a static method (e.g. `ListOffsetRequest.createV1(Map<TopicPartition, Long>)`) since clearly the constructor convention is breaking down?

Internally, it seems it might be easier to convert to `Map<TopicPartition, PartitionData>`. If the version is 1, we can ignore the `maxNumOffsets` field.